File: src/main/java/com/example/tutorial/dsa/medium/simulation/elevator/Elevator.java
package com.example.tutorial.dsa.medium.simulation.elevator;

import java.util.Collections;
import java.util.TreeSet;

/**
 * Individual Elevator class implementing SCAN algorithm.
 *
 * <p><b>SCAN Algorithm (Elevator Algorithm):</b>
 * <ol>
 *   <li>Elevator moves in one direction servicing all requests</li>
 *   <li>When reaching the end, reverses direction</li>
 *   <li>Similar to how a disk arm moves across platters</li>
 * </ol>
 *
 * <p><b>State Tracking:</b>
 * <ul>
 *   <li>currentFloor: where elevator currently is</li>
 *   <li>direction: 1=up, -1=down, 0=idle</li>
 *   <li>upStops: floors to visit while going up (TreeSet for ordering)</li>
 *   <li>downStops: floors to visit while going down (reverse TreeSet)</li>
 * </ul>
 */
public class Elevator {
  private final int id;
  private final int numFloors;
  private int currentFloor;
  private int direction;  // 1=up, -1=down, 0=idle
  private final TreeSet<Integer> upStops;
  private final TreeSet<Integer> downStops;

  public Elevator(int id, int numFloors) {
    this.id = id;
    this.numFloors = numFloors;
    this.currentFloor = 1;
    this.direction = 0;
    this.upStops = new TreeSet<>();
    this.downStops = new TreeSet<>(Collections.reverseOrder());
  }

  /**
   * Calculates score for this elevator to handle a request.
   * Lower score = better choice for assignment.
   *
   * <p><b>Scoring Logic:</b>
   * <ul>
   *   <li>Idle elevator: just distance to floor</li>
   *   <li>Moving towards request in same direction: just distance</li>
   *   <li>Moving away or opposite direction: distance + penalty</li>
   * </ul>
   *
   * @param floor the requested floor
   * @param requestDir requested direction (1=up, -1=down)
   * @return score (lower is better)
   */
  public int getScore(int floor, int requestDir) {
    int distance = Math.abs(currentFloor - floor);

    // Idle elevator - just use distance.
    if (direction == 0) {
      return distance;
    }

    // Moving towards request floor and same direction - best case.
    if (direction == requestDir) {
      if ((direction == 1 && floor >= currentFloor)
          || (direction == -1 && floor <= currentFloor)) {
        return distance;
      }
    }

    // Will need to change direction - add penalty.
    return distance + numFloors;
  }

  /**
   * Adds a stop to this elevator's queue.
   *
   * @param floor the floor to stop at
   */
  public void addStop(int floor) {
    if (floor > currentFloor || (direction == 0 && floor != currentFloor)) {
      upStops.add(floor);
    }
    if (floor < currentFloor || (direction == 0 && floor != currentFloor)) {
      downStops.add(floor);
    }

    // Set direction if idle.
    if (direction == 0) {
      if (!upStops.isEmpty()) {
        direction = 1;
      } else if (!downStops.isEmpty()) {
        direction = -1;
      }
    }
  }

  /**
   * Advances elevator by one floor.
   */
  public void step() {
    if (direction == 0) {
      return;  // Idle, nothing to do.
    }

    // Move one floor in current direction.
    currentFloor += direction;

    // Check if we've arrived at a stop.
    if (direction == 1 && upStops.contains(currentFloor)) {
      upStops.remove(currentFloor);
    } else if (direction == -1 && downStops.contains(currentFloor)) {
      downStops.remove(currentFloor);
    }

    // Check if we need to change direction.
    if (direction == 1 && upStops.isEmpty()) {
      direction = downStops.isEmpty() ? 0 : -1;
    } else if (direction == -1 && downStops.isEmpty()) {
      direction = upStops.isEmpty() ? 0 : 1;
    }
  }

  public int getId() {
    return id;
  }

  public int getCurrentFloor() {
    return currentFloor;
  }

  public int getDirection() {
    return direction;
  }

  public String getDirectionString() {
    return direction == 1 ? "UP" : (direction == -1 ? "DOWN" : "IDLE");
  }

  @Override
  public String toString() {
    return "Elevator " + id + ": Floor " + currentFloor + ", " + getDirectionString()
        + ", upStops=" + upStops + ", downStops=" + downStops;
  }
}

File: src/main/python/com/example/tutorial/dsa/medium/simulation/elevator/elevator.py
"""
Individual Elevator class implementing SCAN algorithm.

SCAN Algorithm (Elevator Algorithm):
    1. Elevator moves in one direction servicing all requests
    2. When reaching the end, reverses direction
    3. Similar to how a disk arm moves across platters

State Tracking:
    - current_floor: where elevator currently is
    - direction: 1=up, -1=down, 0=idle
    - up_stops: floors to visit while going up
    - down_stops: floors to visit while going down
"""

from typing import Set


class Elevator:
    """Individual elevator implementing SCAN algorithm."""

    def __init__(self, elevator_id: int, num_floors: int):
        self.id = elevator_id
        self.num_floors = num_floors
        self.current_floor = 1
        self.direction = 0  # 1=up, -1=down, 0=idle
        self.up_stops: Set[int] = set()
        self.down_stops: Set[int] = set()

    def get_score(self, floor: int, request_dir: int) -> int:
        """
        Calculates score for this elevator to handle a request.
        Lower score = better choice for assignment.

        Scoring Logic:
            - Idle elevator: just distance to floor
            - Moving towards request in same direction: just distance
            - Moving away or opposite direction: distance + penalty

        Args:
            floor: the requested floor
            request_dir: requested direction (1=up, -1=down)

        Returns:
            score (lower is better)
        """
        distance = abs(self.current_floor - floor)

        # Idle elevator - just use distance.
        if self.direction == 0:
            return distance

        # Moving towards request floor and same direction - best case.
        if self.direction == request_dir:
            if (self.direction == 1 and floor >= self.current_floor) or \
               (self.direction == -1 and floor <= self.current_floor):
                return distance

        # Will need to change direction - add penalty.
        return distance + self.num_floors

    def add_stop(self, floor: int) -> None:
        """
        Adds a stop to this elevator's queue.

        Args:
            floor: the floor to stop at
        """
        if floor > self.current_floor or (self.direction == 0 and floor != self.current_floor):
            self.up_stops.add(floor)
        if floor < self.current_floor or (self.direction == 0 and floor != self.current_floor):
            self.down_stops.add(floor)

        # Set direction if idle.
        if self.direction == 0:
            if self.up_stops:
                self.direction = 1
            elif self.down_stops:
                self.direction = -1

    def step(self) -> None:
        """Advances elevator by one floor."""
        if self.direction == 0:
            return  # Idle, nothing to do.

        # Move one floor in current direction.
        self.current_floor += self.direction

        # Check if we've arrived at a stop.
        if self.direction == 1 and self.current_floor in self.up_stops:
            self.up_stops.remove(self.current_floor)
        elif self.direction == -1 and self.current_floor in self.down_stops:
            self.down_stops.remove(self.current_floor)

        # Check if we need to change direction.
        if self.direction == 1 and not self.up_stops:
            self.direction = 0 if not self.down_stops else -1
        elif self.direction == -1 and not self.down_stops:
            self.direction = 0 if not self.up_stops else 1

    def get_direction_string(self) -> str:
        """Returns direction as string."""
        return "UP" if self.direction == 1 else ("DOWN" if self.direction == -1 else "IDLE")

    def __str__(self) -> str:
        return (f"Elevator {self.id}: Floor {self.current_floor}, {self.get_direction_string()}, "
                f"upStops={sorted(self.up_stops)}, downStops={sorted(self.down_stops, reverse=True)}")
